tests/integration.rs

```diff
@@ -262,117 +262,164 @@ fn time_constructor() {
+                "signup.netflix.com",
+                "www.netflix.ca",
+                "www1.netflix.com",
+                "www2.netflix.com",
+                "www3.netflix.com",
+                "develop-stage.netflix.com",
+                "release-stage.netflix.com",
+                // NOT 'www.netflix:com'
+            ],
+        );
+    }
+
+    #[test]
+    pub fn wildcard_subject_alternative_names() {
+        // same as netflix ee certificate, but with the last name in the list
+        // changed to 'ww*.netflix:com'
+        let data = include_bytes!("misc/dns_names_and_wildcards.der");
+
+        expect_cert_dns_names(
+            data,
+            &[
+                "account.netflix.com",
+                "*.netflix.com",
+                "netflix.ca",
+                "netflix.com",
+                "signup.netflix.com",
+                "www.netflix.ca",
+                "www1.netflix.com",
+                "www2.netflix.com",
+                "www3.netflix.com",
+                "develop-stage.netflix.com",
+                "release-stage.netflix.com",
+                "www.netflix.com",
+            ],
+        );
+    }
+
+    #[test]
+    pub fn no_subject_alt_names() {
+        let ee = CertificateDer::from(&include_bytes!("misc/no_subject_alternative_name.der")[..]);
+
+        let cert = webpki::EndEntityCert::try_from(&ee)
+            .expect("should parse end entity certificate correctly");
+
+        let names = cert
+            .dns_names()
+            .expect("we should get a result even without subjectAltNames");
+
+        assert!(names.collect::<Vec<_>>().is_empty());
+    }
+
+    // This test reproduces (part of)
+    // https://github.com/rustls/webpki/issues/167 --- an end-entity cert where
+    // the common name is a `PrintableString` rather than a `UTF8String` cannot
+    // iterate over its subject alternative names.
+    #[test]
+    pub fn printable_string_common_name() {
+        const DNS_NAME: &str = "test.example.com";
+
+        let alg = &rcgen::PKCS_ECDSA_P256_SHA256;
+
+        let issuer = mk_issuer(alg);
+
+        let ee_cert_der = {
+            let mut params = rcgen::CertificateParams::new(vec![DNS_NAME.to_string()]);
+            params.distinguished_name.push(
+                rcgen::DnType::CommonName,
+                rcgen::DnValue::PrintableString("example.com".to_string()),
+            );
+            params.is_ca = rcgen::IsCa::ExplicitNoCa;
+            params.alg = alg;
+            let cert = rcgen::Certificate::from_params(params)
+                .expect("failed to make ee cert (this is a test bug)");
+            let der = cert
+                .serialize_der_with_signer(&issuer)
+                .expect("failed to serialize signed ee cert (this is a test bug)");
+            der
+        };
+
+        expect_cert_dns_names(&ee_cert_derd, &[DNS_NAME]);
+    }
+
+    fn mk_issuer(alg: &'static rcgen::SignatureAlgorithm) -> rcgen::Certificate {
```

Comment:
This helper is pretty close to the `verify_cert.rs` helper `make_issuer`. 

Is there a convenient way to allow both `tests/integration.rs` and `src/verify_cert.rs`'s unit tests to share the same rcgen helpers?

Reply:
Hmm. Unfortunately, I think the only way to share the helper between tests inside the library and integration tests is to make it a public API, which we probably don't want to do. Alternatively, I could move these tests out of integration and into `src/verify_cert.rs`, but they felt more like integration tests because they just interact with public APIs and I wanted to be able to use the existing `expect_cert_dns_names` helper in the new tests. Would you prefer them in `verify_cert.rs`? I'm happy to move them.

Reply:
> Hmm. Unfortunately, I think the only way to share the helper between tests inside the library and integration tests is to make it a public API, which we probably don't want to do.

Blech, yeah that's no good. 

>  Alternatively, I could move these tests out of integration and into src/verify_cert.rs, but they felt more like integration tests because they just interact with public APIs and I wanted to be able to use the existing expect_cert_dns_names helper in the new tests. Would you prefer them in verify_cert.rs? I'm happy to move them.

I think `src/end_entity.rs` might be a better fit since the unit tests are specific to the `EndEntity::dns_names` fn and not related to certificate validation, but that shifts the problem to how to have `src/end_entity.rs` and `src/verify_cert.rs` share helpers. :thinking: However, maybe that's more more tractable there since they're both crate-internal locations?

I don't have a super strong preference.

@djc @ctz Do you have thoughts? Another idea is to leave the duplication in and consider upstream work in `rcgen` that could reduce the need for the helper boilerplate. I have some ideas there but it will take some time to implement.

Reply:
We can more easily share test helpers within the crate. I think moving these tests to `end_entity` and making a shared crate-private `test_util` module would be the best solution here, probably.

Reply:
>  making a shared crate-private test_util module would be the best solution here, probably.

SGTM. Let's roll with that and resolve this conversation.

Reply:
I think for sharing test-only code between unit and integration tests, it's reasonable to put that code in the crate, and then reuse it from integration tests by

```rust
#[path = "../src/test_util.rs"]
mod test_util;
```

That means the code in `test_util.rs` is more restricted than usual (it can't reference crate internals, etc.) But for helpers and test fixtures that should be OK.